### DISCUSSION TITLE
Compiling javascript before running system specs

### DISCUSSION THREAD

**Discussion by tonekk** [rspec/rspec-rails#2709] [Q&A]
Hi community,

not sure if this is a good place to ask, but maybe you can point me to the right one if this isn't it.

I'm using [jsbundling-rails](jsbundling-rails) in conjunction with capybara and [ferrum](https://github.com/rubycdp/ferrum) for my system specs. I noticed that using this setup, changes to javascript are not compiled automatically when executing system specs, which is something that I was used to back when I was using webpacker.

I ended up putting this into my `rails_helper.rb`:
```ruby
  # Precompile the assets (used by system tests)
  config.before :all, type: :system do
    if !ENV["GITHUB_ACTIONS"] && !ENV["ASSET_PRECOMPILE_DONE"]
      puts "\n\nPrecompiling assets for system tests..."
      prep_passed = system "rake assets:precompile"
      ENV["ASSET_PRECOMPILE_DONE"] = "true"
      abort "\nYour assets didn't compile. Exiting WITHOUT running any tests. Review the output above to resolve any errors." if !prep_passed
    end
  end
```

Seemed like a straightforward solution, but it seems that rspec caches(?) the existing javascript on boot-up, so I have to execute it twice in order for the javascript to be up-to-date.
I ended up modifying my binstub to compile the js files before rspec even starts:
```ruby
#!/usr/bin/env ruby
ENV["BUNDLE_GEMFILE"] ||= File.expand_path("../Gemfile", __dir__)

# Determine whether to compile assets
should_compile_assets = ARGV.empty? || ARGV.any? { |arg| arg.include?("/system/") }

if should_compile_assets
  output = `bundle exec rake assets:precompile 2>&1`
  abort(output) unless $?.success?
end
```

This works, but it's a little inelegant. Is there another way?
I'm curious if there is anyway to prevent rspec from caching the js files before I'm bundling them.

Thanks!

**Comment by georgialive** [comment]
Your question is indeed relevant for Rails development, especially when transitioning from Webpacker to jsbundling-rails. The issue you're facing with RSpec and Capybara not automatically picking up changes in JavaScript files is a common one in this scenario. Here are a few suggestions and alternative approaches to consider:

### Understanding the Problem
- **Caching Issue**: RSpec, when used with system tests, doesn't automatically detect changes in the asset pipeline, which includes JavaScript files managed by jsbundling-rails. This behavior differs from Webpacker, which had its own dev server to handle such updates.

- **Precompiling Assets**: Your approach to precompile assets before running tests ensures that the latest version of JavaScript files is used. However, it adds overhead and may slow down the test suite, especially if assets are precompiled for every test run.

### Alternative Approaches

1. **Conditional Precompilation**: You're on the right track with conditional precompilation. Enhancing this approach might involve checking the timestamp of JavaScript files versus the compiled assets and only recompiling if there are changes. This can reduce unnecessary compilations.

2. **Listening for File Changes**: Consider using file watchers that trigger asset recompilation when JavaScript files change. Tools like `guard` or `listen` in Ruby can be configured to watch for file changes and execute commands accordingly.

3. **Integration with Test Environment**: You can integrate the asset compilation more closely with your test environment setup. For instance, you could configure RSpec or Rails to check if asset compilation is needed every time the test environment is loaded.

4. **Improving the RSpec Configuration**: Modify your `rails_helper.rb` or `spec_helper.rb` to better handle the compilation process. This could include more intelligent checks for when assets need to be recompiled.

5. **Leveraging Docker or CI Tools**: If you're using Docker or CI tools like GitHub Actions, you can set up your environment to handle these compilations more efficiently as part of the setup process.

6. **Reducing Asset Size**: If the asset compilation time is a significant concern, consider reducing the size of your assets or splitting them into smaller chunks that can be compiled more quickly.

7. **Community Feedback**: Since this is a relatively common issue, reaching out to the Rails community (such as on forums, Stack Overflow, or Rails GitHub issues) can provide more insights or shared experiences on how others have tackled this issue.

### Addressing Your Specific Setup
- **Inefficiency in Current Approach**: As you noted, your current setup, while functional, is a bit inelegant and potentially slow. Optimizing it for better efficiency without sacrificing the accuracy of your tests should be the goal.

- **Exploring Rails and RSpec Updates**: Ensure you're using the latest versions of Rails, RSpec, and jsbundling-rails, as newer versions might offer better integration or solutions to this issue.

While there might not be a perfect solution, exploring these alternatives and combinations thereof can lead to a more efficient and elegant setup for your Rails application testing environment.

**Reply by tonekk** [reply]
Hello @georgialive ,

I hope this message finds you well. I wanted to extend my sincere thanks for the time and effort you put into responding to my query on GitHub. Your detailed answer was not only insightful but also showcased a deep understanding of the subject. I truly appreciate your willingness to help and the depth of knowledge you shared.

While your response was thorough and informative, I realize I may not have articulated my question clearly enough, leading to a slight misalignment with what I was seeking. My apologies for any confusion caused. To clarify, my question was more specifically focused on [briefly describe the specific aspect or angle of your question that was missed].

I believe your expertise could provide valuable insights into this particular aspect, and I would be grateful for any further thoughts you might have on this matter.

Thank you once again for your contribution and for being a part of the community. Your willingness to assist others is truly commendable.

Warm regards,

@tonekk

**Comment by pirj** [comment]
Related #2675

**Comment by JonRowe** [comment]
For the record RSpec does no compilation of assets directly, what ever behaviour exists is brought in from Rails test helpers, if theres something more we can bring in to get parity with MiniTest I'm happy to review a PR, but this could also be an ordering issue where something isn't being included in the right place
